postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+    }    
+
+};
+
+
+/*****************************************************************************/
+/* POSTGRESQL IMPORT PROCEDURE                                               */
+/*****************************************************************************/
+
+struct PostgresqlImportConfig: public ProcedureConfig {
+
+    static constexpr const char * name = "postgresql.import";
+
+    string databaseName;
+    int port;
+    string host;
+    string postgresqlQuery;
+
+    /// The output dataset.  Rows will be dumped into here via insertRows.
+    PolyConfigT<Dataset> outputDataset;
+
+    PostgresqlImportConfig() {
+        databaseName = "database";
+        port = 1234;
+        host = "localhost";
+        postgresqlQuery = "";
+
+        outputDataset.withType("sparse.mutable");
+    }
+
+};
+
+DECLARE_STRUCTURE_DESCRIPTION(PostgresqlImportConfig);
+DEFINE_STRUCTURE_DESCRIPTION(PostgresqlImportConfig);
+
+PostgresqlImportConfigDescription::
+PostgresqlImportConfigDescription()
+{
+    addParent<ProcedureConfig>();
+
+    addField("databaseName", &PostgresqlImportConfig::databaseName, "Name of the database to connect to.");
+    addField("port", &PostgresqlImportConfig::port, "Port of the database to connect to.", 1234);
+    addField("host", &PostgresqlImportConfig::host, "Address of the database to connect to ");
+    addField("postgresqlQuery", &PostgresqlImportConfig::postgresqlQuery, "Query to run in postgresql to get rows");
+
+    addField("outputDataset", &PostgresqlImportConfig::outputDataset,
+             "Output dataset configuration.  This may refer either to an "
+             "existing dataset, or a fully specified but non-existing dataset "
+             "which will be created by the procedure.", PolyConfigT<Dataset>().withType("sparse.mutable"));
+}
+
+struct PostgresqlImportProcedure: public Procedure {
+
+    PostgresqlImportConfig procedureConfig;
+
+    PostgresqlImportProcedure(MldbServer * server,
+                         const PolyConfig & config,
+                         std::function<bool (Json::Value)> onProgress)
+        : Procedure(server)
+    {
+        procedureConfig = config.params.convert<PostgresqlImportConfig>();
+    }
+
+    pg_conn* startConnection(const PostgresqlImportConfig & config_) const
+    {
+        return startPostgresqlConnection(config_.databaseName, config_.host, config_.port);
+    }
+
+    virtual Any getStatus() const override
+    {
+        Json::Value result;
+        result["ok"] = true;
+        return result;
+    }
+
+    virtual RunOutput run(const ProcedureRunConfig & run,
+                          const std::function<bool (const Json::Value &)> & onProgress) const override
+    {        
+        RunOutput result;
+
+        auto runProcConf = applyRunConfOverProcConf(procedureConfig, run);
+
+        // Connect to Postgresl database
+        auto conn = startConnection(runProcConf);
```

Comment:
Consider using an object that will close the `conn` and one that will clear `res` once they go out of scope.